docs/app-rails/application-security.md

```diff
@@ -0,0 +1,136 @@
+#### 2.5 Replay Attacks for CookieStore Sessions
+1. **TODO:** We’re not using a nonce generator that would protect against cookie replay attacks. The commented out code for this is located in `/app-rails/config/initializers/generate_security_policy.rb` This however ties users to a particular server/session as we don’t want to save nonces in the database.
+#### 2.6 Session Fixation
+#### 2.7 Session Fixation - Countermeasures
+1. Devise will automatically expire sessions on sign in and sign out with `config.expire_all_remember_me_on_sign_out = true`.
+#### 2.8 Session Expiry
+
+### 3 Cross-Site Request Forgery (CSRF)
+#### 3.1  CSRF Countermeasures
+1. GET, POST, DELETE, and rails’ resources are used appropriately in the `routes.rb` file.
+1. **TODO:** We’re not setting forgery protection in production despite passing a CSRF token to the client, therefore If I change my CSRF token manually in my browser, I’m not prevented from taking any actions (GET or POST requests) as a signed in user.
+
+### 4 Redirection and Files
+#### 4.1 Redirection
+1. link_tos do not route to user inputs, which is XSS vulnerable.
+1. redirect_to does not use params or user inputs in the routes.
+4.2 File Uploads
+1. **TODO:** App doesn’t currently handle file imports but that should be documented as it is a common need in Government Applications.
+
+### 5. User Management
+1. When using devise we don’t need to set has_secure_password on password fields.
+#### 5.1 Brute Forcing Accounts
+1. Username error is generic and does not indicate whether it was an error with the username or password.
+1. Forgot password confirms the email was sent, and not whether the username exists.
+#### 5.2 Account Hijacking
+1. Change password includes email 6 digit code.
+1. **TODO:** Require entering the password when changing email.
+#### 5.3 CAPTCHA
+1. **TODO:** Include Captcha on account creation, login, change password, and change email.
+1. **TODO:** Include honeypot fields and logic on Non logged in forms to catch bots that spam all fields (good resource: https://nedbatchelder.com/text/stopbots.html).
+#### 5.4 Logging
+1. We’re filtering the following parameter partials (if all or any part matches) in the `config/initialization/filter_paramater_logging.rb` file: `:passw, :secret, :token, :_key, :crypt, :salt, :certificate, :otp, :ssn`.
+#### 5.5 Regex
+1. Two places where regex is used (`mailto.rb` and devise) use the correct ruby `\A` and `\z` and not the more common: `/^` and `$/`.
+1. **TODO:** Add multiline: true to our regex format: in validations.
+#### 5.6 Privilege Escalation
+1. **TODO:** By manually changing a parameter, a hacker may get access to information they should not be able to access. Instead of doing things like: `@task = Task.find(params[:id])`, instead do `@user.tasks.find(params[:id])`. This is not in the template, but it is in the app the template is based on.
```

Comment:
This is a good catch, but does not belong in this document because this is for the template. Please create an issue for this in the PFML Accelerator repo.

Reply:
I looked into this a bit more. In general the application is using a library, [pundit](https://github.com/varvet/pundit), to wire up resource authorization and is calling the authorize method when looking up models. It seems like a good best practice to look up the model from and id based on the user, but it may be redundant to create tickets for that if we're already using Pundit. Thoughts?

In the specific case I found, it's in the Admin view, so I'm guessing that should be allowed. However, it should also add the `authorize` method it's using elsewhere in case we either don't want admins to be able to see any data from any user, or if we want to have different types of admins with different levels of permissions. But maybe that doesn't need to happen at this point?